### PR TITLE
feat: add bacnet stack

### PR DESCRIPTION
**Pull request by gocarlos**
Specify library name and version:  **bacnet-stack/master**

bacnet-stack got cross platform support into master but still no release available so far

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/wiki#how-to-submit-a-pull-request) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by gocarlos** [comment]
related: https://github.com/bacnet-stack/bacnet-stack/issues/48

File: recipes/bacnet-stack/all/conandata.yml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+sources:
+  "677f528aa46bebe5b7a129342b07fc491acd32b1":
```
Comment (madebr) [inline recipes/bacnet-stack/all/conandata.yml]:
```suggestion
  "20200218":
```
The policy of this repo is to use the date when the commit is made.
Actually, it is to use the latest commit on the specified date.

File: recipes/bacnet-stack/all/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    def config_options(self):
+        if self.settings.os == "Windows":
+            del self.options.fPIC
+
+    def configure(self):
+        del self.settings.compiler.libcxx
+        del self.settings.compiler.cppstd
+
+    def source(self):
+
```
Comment (madebr) [inline recipes/bacnet-stack/all/conanfile.py]:
```suggestion
```

File: recipes/bacnet-stack/all/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        del self.settings.compiler.cppstd
+
+    def source(self):
+        
+        tools.get(**self.conan_data["sources"][self.version])
+        extracted_dir = self.name + "-" + self.version
+        os.rename(extracted_dir, self._source_subfolder)
+
+    def _configure_cmake(self):
+        cmake = CMake(self)
```
Comment (madebr) [inline recipes/bacnet-stack/all/conanfile.py]: Cache the `cmake` object so reconfiguration is not needed in `build` and `package`.
```suggestion
        if self._cmake:
            return self._cmake
        self._cmake = CMake(self)
```

File: recipes/bacnet-stack/all/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+        cmake.definitions["BACNET_STACK_BUILD_APPS"] = False
```
Comment (madebr) [inline recipes/bacnet-stack/all/conanfile.py]:
```suggestion
        self._cmake.definitions["BACNET_STACK_BUILD_APPS"] = False
```

File: recipes/bacnet-stack/all/conanfile.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        cmake.configure()
```
Comment (madebr) [inline recipes/bacnet-stack/all/conanfile.py]:
```suggestion
        self._cmake.configure(build_folder=self._build_subfolder)
```

File: recipes/bacnet-stack/all/conanfile.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+        return cmake
```
Comment (madebr) [inline recipes/bacnet-stack/all/conanfile.py]:
```suggestion
        return self._cmake
```

File: recipes/bacnet-stack/all/conanfile.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    default_options = {
+        "shared": False,
+        "fPIC": True
+    }
+    generators = "cmake", "cmake_find_package"
+
+    @property
+    def _source_subfolder(self):
+        return "source_subfolder"
+
```
Comment (madebr) [inline recipes/bacnet-stack/all/conanfile.py]:
```suggestion

    @property
    def _build_subfolder(self):
        return "build_subfolder"

File: recipes/bacnet-stack/all/conanfile.py [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+        cmake = self._configure_cmake()
+        cmake.install()
+        tools.rmdir(os.path.join(self.package_folder, "lib", "cmake"))
+        tools.rmdir(os.path.join(self.package_folder,
+                                 "lib", "bacnet-stack", "cmake"))
+
+    def package_info(self):
+        self.cpp_info.libs = tools.collect_libs(self)
+        if tools.os_info.is_linux:
+            self.cpp_info.libs.append("pthread")
```
Comment (madebr) [inline recipes/bacnet-stack/all/conanfile.py]:
```suggestion
            self.cpp_info.system_libs = ["pthread"]
```

File: recipes/bacnet-stack/all/test_package/conanfile.py [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+class TestConan(ConanFile):
+    settings = "os", "compiler", "build_type", "arch"
+    generators = "cmake"
+
+    def build(self):
+        cmake = CMake(self)
+        cmake.configure()
+        cmake.build()
+
+    def imports(self):
```
Comment (madebr) [inline recipes/bacnet-stack/all/test_package/conanfile.py]: Copying these libraries should not be necessary as conan will handle the dependencies.
```suggestion
```

File: recipes/bacnet-stack/all/test_package/conanfile.py [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+        self.copy("*.dll", dst="bin", src="bin")
```
Comment (madebr) [inline recipes/bacnet-stack/all/test_package/conanfile.py]:
```suggestion
```

File: recipes/bacnet-stack/all/test_package/conanfile.py [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+        self.copy("*.dylib*", dst="bin", src="lib")
```
Comment (madebr) [inline recipes/bacnet-stack/all/test_package/conanfile.py]:
```suggestion
```

File: recipes/bacnet-stack/all/test_package/conanfile.py [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+        self.copy('*.so*', dst='bin', src='lib')
```
Comment (madebr) [inline recipes/bacnet-stack/all/test_package/conanfile.py]:
```suggestion
```

File: recipes/bacnet-stack/all/test_package/conanfile.py [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+
```
Comment (madebr) [inline recipes/bacnet-stack/all/test_package/conanfile.py]:
```suggestion
```

**Comment by madebr** [comment]
@gocarlos When you click on `Files changed` on top of this page, you can combine proposed changes.

File: recipes/bacnet-stack/all/test_package/conanfile.py [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+        cmake = CMake(self)
+        cmake.configure()
+        cmake.build()
+
+        self.copy("*.dll", dst="bin", src="bin")
+        self.copy("*.dylib*", dst="bin", src="lib")
+        self.copy('*.so*', dst='bin', src='lib')
+    def test(self):
+        if not tools.cross_building(self.settings):
+            os.chdir("bin")
```
Comment (madebr) [inline recipes/bacnet-stack/all/test_package/conanfile.py]:
```suggestion
            bin_path = os.path.join("bin", "test_package")
```

File: recipes/bacnet-stack/all/test_package/conanfile.py [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+            self.run(".%sexample" % os.sep)
```
Comment (madebr) [inline recipes/bacnet-stack/all/test_package/conanfile.py]:
```suggestion
            bin_path = self.run(bin_path, run_environment=True)
```

File: recipes/bacnet-stack/config.yml [hunk 16]
```diff
@@ -0,0 +1,4 @@
+---
+versions:
+  "bacnet-stack-0.8.6":
```
Comment (madebr) [inline recipes/bacnet-stack/config.yml]:
```suggestion
  "0.8.6":
```

File: recipes/bacnet-stack/all/conanfile.py [hunk 17]
```diff
@@ -0,0 +1,66 @@
+import os
+from conans import CMake, ConanFile, tools
+from conans.errors import ConanInvalidConfiguration
+
+
+class BacnetStackConan(ConanFile):
+    name = "bacnet-stack"
+    license = "BSD"
```
Comment (uilianries) [inline recipes/bacnet-stack/all/conanfile.py]:
```suggestion
    license = "GPL-2.0-or-later"
```

File: recipes/bacnet-stack/all/conanfile.py [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+        cmake.definitions["BACNET_STACK_BUILD_APPS"] = False
+        cmake.configure()
+        return cmake
+
+    def build(self):
+        cmake = self._configure_cmake()
+        cmake.build()
+
+    def package(self):
+        self.copy("LICENSE", dst='licenses', src=self._source_subfolder)
```
Comment (uilianries) [inline recipes/bacnet-stack/all/conanfile.py]:
```suggestion
        self.copy("gpl-2.txt", dst='licenses', src=os.path.join(elf._source_subfolder, "license"))
```

File: recipes/bacnet-stack/config.yml (hunk 16 above)
Comment (gocarlos) [reply recipes/bacnet-stack/config.yml]: This is the tag name

File: recipes/bacnet-stack/config.yml [hunk 19]
```diff
@@ -0,0 +1,4 @@
+---
+versions:
+  "0.8.6":
```
Comment (madebr) [inline recipes/bacnet-stack/config.yml]:
```suggestion
  "20200218":
```
The version inside the `config.yml` and `conandata.yml` must match.

File: recipes/bacnet-stack/all/conandata.yml [hunk 20]
```diff
@@ -0,0 +1,4 @@
+sources:
+  "20200218":
```
Comment (madebr) [inline recipes/bacnet-stack/all/conandata.yml]:
```suggestion
  "0.8.6":
```

File: recipes/bacnet-stack/all/conandata.yml [hunk 21]
```diff
@@ -0,0 +1,4 @@
...
+    sha256: 1fee2f943c2d3e2cbb3a2f9ff3de40b392418c96c751123c054530cc29506b3a
+    url: https://github.com/bacnet-stack/bacnet-stack/archive/677f528aa46bebe5b7a129342b07fc491acd32b1.tar.gz
```
Comment (madebr) [inline recipes/bacnet-stack/all/conandata.yml]:
```suggestion
    url: https://github.com/bacnet-stack/bacnet-stack/archive/bacnet-stack-0.8.6.tar.gz
```

File: recipes/bacnet-stack/all/conandata.yml [hunk 22]
```diff
@@ -0,0 +1,4 @@
+sources:
+  "20200218":
+    sha256: 1fee2f943c2d3e2cbb3a2f9ff3de40b392418c96c751123c054530cc29506b3a
```
Comment (madebr) [inline recipes/bacnet-stack/all/conandata.yml]:
```suggestion
    sha256: 7ea0d145164d30ecfebd97dce02a2347de88afb2458b144b20f8fc44c2bb41f4
```

File: recipes/bacnet-stack/all/conandata.yml (hunk 21 above)
Comment (gocarlos) [reply recipes/bacnet-stack/all/conandata.yml]: I guess this wont work as this version is 2 years old and has no cmake support

Comment (gocarlos) [reply recipes/bacnet-stack/all/conandata.yml]: Nor class platform support

Comment (madebr) [reply recipes/bacnet-stack/all/conandata.yml]: Ah, good point. Then disregard this suggestion!

File: recipes/bacnet-stack/config.yml (hunk 19 above)
Comment (madebr) [inline recipes/bacnet-stack/config.yml]:
```suggestion
  "20200218":
```

File: recipes/bacnet-stack/all/conanfile.py [hunk 23]
```diff
@@ -0,0 +1,65 @@
+import os
+from conans import CMake, ConanFile, tools
+from conans.errors import ConanInvalidConfiguration
+
+
+class BacnetStackConan(ConanFile):
+    name = "bacnet-stack"
+    license = "GPL-2.0-or-later"
+    url = "https://github.com/bacnet-stack/bacnet-stack/"
```
Comment (madebr) [inline recipes/bacnet-stack/all/conanfile.py]:
```suggestion
    homepage = "https://github.com/bacnet-stack/bacnet-stack/"
    url = "https://github.com/conan-io/conan-center-index"
```

File: recipes/bacnet-stack/all/conanfile.py [hunk 24]
```diff
@@ -0,0 +1,65 @@
+import os
+from conans import CMake, ConanFile, tools
+from conans.errors import ConanInvalidConfiguration
```
Comment (madebr) [inline recipes/bacnet-stack/all/conanfile.py]:
```suggestion
```

File: recipes/bacnet-stack/all/conanfile.py [hunk 25]
```diff
@@ -0,0 +1,68 @@
...
+        if self.settings.os == "Windows":
+            del self.options.fPIC
+
+    def configure(self):
+        del self.settings.compiler.libcxx
+        del self.settings.compiler.cppstd
+
+    def source(self):
+        tools.get(**self.conan_data["sources"][self.version])
+        extracted_dir = self.name + "-" + self.version
```
Comment (madebr) [inline recipes/bacnet-stack/all/conanfile.py]: See `https://github.com/conan-io/conan-center-index/pull/784#discussion_r378786929`
```suggestion
        extracted_dir = os.path.splitext(os.path.basename(self.conan_data["sources"][self.version]["url"]))[0]
```

File: recipes/bacnet-stack/all/conanfile.py [hunk 26]
```diff
@@ -0,0 +1,68 @@
...
+    generators = "cmake", "cmake_find_package"
+
+    @property
+    def _source_subfolder(self):
+        return "source_subfolder"
+
+    @property
+    def _build_subfolder(self):
+        return "build_subfolder"
+    def config_options(self):
```
Comment (madebr) [inline recipes/bacnet-stack/all/conanfile.py]:
```suggestion

    def config_options(self):
```

File: recipes/bacnet-stack/all/conanfile.py [hunk 27]
```diff
@@ -0,0 +1,68 @@
...
+        if self.settings.os == "Windows":
+            del self.options.fPIC
+
+    def configure(self):
+        del self.settings.compiler.libcxx
+        del self.settings.compiler.cppstd
+
+    def source(self):
+        tools.get(**self.conan_data["sources"][self.version])
+        extracted_dir = os.path.splitext(os.path.basename(self.conan_data["sources"][self.version]["url"]))[0]
```
Comment (madebr) [inline recipes/bacnet-stack/all/conanfile.py]: Apparently, this does not work on a `.tar.gz`
```suggestion
        extracted_dir = os.path.basename(self.conan_data["sources"][self.version]["url"]).split(".")[0]
```

File: recipes/bacnet-stack/all/conanfile.py [hunk 28]
```diff
@@ -0,0 +1,69 @@
...
+        if self.settings.os == "Windows":
+            del self.options.fPIC
+
+    def configure(self):
+        del self.settings.compiler.libcxx
+        del self.settings.compiler.cppstd
+
+    def source(self):
+        tools.get(**self.conan_data["sources"][self.version])
+        extracted_dir = os.path.basename(self.conan_data["sources"][self.version]["url"]).split(".")[0]
```
Comment (madebr) [inline recipes/bacnet-stack/all/conanfile.py]: Sorry!!
```suggestion
        extracted_dir = self.name + "-" + os.path.basename(self.conan_data["sources"][self.version]["url"]).split(".")[0]
```

**Comment by Croydon** [comment]
Date-based versions were only meant for cases when there aren't any releases, not for cases in which there are releases, but not a new one happened in a longer time

This would mean that for this package in the future, date-based versions and regular versions numbers would be mixed which is pretty confusing

**Comment by madebr** [comment]
@Croydon What would you suggest in this case?
Wait for upstream for a new release?

File: recipes/bacnet-stack/all/conanfile.py [hunk 29]
```diff
@@ -0,0 +1,69 @@
...
+    options = {
+        "shared": [True, False],
+        "fPIC": [True, False],
+    }
+    default_options = {
+        "shared": False,
+        "fPIC": True
+    }
+    generators = "cmake", "cmake_find_package"
+
```
Comment (madebr) [inline recipes/bacnet-stack/all/conanfile.py]:
```suggestion

    _cmake = None

```

File: recipes/bacnet-stack/all/CMakeLists.txt [hunk 30]
```diff
@@ -0,0 +1,7 @@
+cmake_minimum_required(VERSION 2.8.11)
+project(cmake_wrapper)
+
+include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
```
Comment (madebr) [inline recipes/bacnet-stack/all/CMakeLists.txt]:
```suggestion
include(conanbuildinfo.cmake)
```

File: recipes/bacnet-stack/all/CMakeLists.txt [hunk 31]
```diff
@@ -0,0 +1,7 @@
...
+conan_basic_setup()
+
+add_subdirectory("source_subfolder")
```
Comment (madebr) [inline recipes/bacnet-stack/all/CMakeLists.txt]:
```suggestion
add_subdirectory(source_subfolder)
```

File: recipes/bacnet-stack/all/conanfile.py [hunk 32]
```diff
@@ -0,0 +1,71 @@
...
+class BacnetStackConan(ConanFile):
+    name = "bacnet-stack"
+    license = "GPL-2.0-or-later"
+    homepage = "https://github.com/bacnet-stack/bacnet-stack/"
+    url = "https://github.com/conan-io/conan-center-index"
+    description = """
+        BACnet Protocol Stack library provides a BACnet application layer,
+        network layer and media access (MAC) layer communications services."""
+    topics = ("bacnet")
+    settings = "os", "compiler", "build_type", "arch"
```
Comment (madebr) [inline recipes/bacnet-stack/all/conanfile.py]:
```suggestion
    exports_sources = "CMakeLists.txt"
    settings = "os", "compiler", "build_type", "arch"
```

File: recipes/bacnet-stack/all/conanfile.py [hunk 33]
```diff
@@ -0,0 +1,71 @@
...
+        self._cmake.definitions["BACNET_STACK_BUILD_APPS"] = False
+        self._cmake.configure(build_folder=self._build_subfolder)
+        return self._cmake
+
+    def build(self):
+        cmake = self._configure_cmake()
+        cmake.build()
+
+    def package(self):
+        self.copy("gpl-2.txt", dst='licenses', src=os.path.join(elf._source_subfolder, "license"))
```
Comment (madebr) [inline recipes/bacnet-stack/all/conanfile.py]:
```suggestion
        self.copy("gpl-2.txt", dst='licenses', src=os.path.join(self._source_subfolder, "license"))
```

**Comment by madebr** [comment]
@gocarlos 
I think `bacnet-stack`'s cmake build system is not yet mature.
Building a shared library fails because some source files are not included in the build.
Also, some files require openssl installed and the cmake script is not looking for openssl.
These are problems that should be fixed in the upstream project, not in CCI as a patch.

I suggest adding shared cmake builds to the ci to catch these errors and fixing them later.

**Comment by madebr** [comment]
@gocarlos  I opened https://github.com/bacnet-stack/bacnet-stack/issues/49
So upstream can take a look at the problem.

File: recipes/bacnet-stack/all/conanfile.py [hunk 34]
```diff
@@ -0,0 +1,72 @@
...
+        self.copy("gpl-2.txt", dst='licenses', src=os.path.join(self._source_subfolder, "license"))
+        cmake = self._configure_cmake()
+        cmake.install()
+        tools.rmdir(os.path.join(self.package_folder, "lib", "cmake"))
+        tools.rmdir(os.path.join(self.package_folder,
+                                 "lib", "bacnet-stack", "cmake"))
+
+    def package_info(self):
+        self.cpp_info.libs = tools.collect_libs(self)
+        if tools.os_info.is_linux:
```
Comment (madebr) [inline recipes/bacnet-stack/all/conanfile.py]:
```suggestion
        if self.settings.os == "Linux":
```

File: recipes/bacnet-stack/all/conanfile.py [hunk 35]
```diff
@@ -0,0 +1,72 @@
...
+    def package(self):
+        self.copy("gpl-2.txt", dst='licenses', src=os.path.join(self._source_subfolder, "license"))
+        cmake = self._configure_cmake()
+        cmake.install()
+        tools.rmdir(os.path.join(self.package_folder, "lib", "cmake"))
+        tools.rmdir(os.path.join(self.package_folder,
+                                 "lib", "bacnet-stack", "cmake"))
+
+    def package_info(self):
+        self.cpp_info.libs = tools.collect_libs(self)
```
Comment (madebr) [inline recipes/bacnet-stack/all/conanfile.py]:
```suggestion
        self.cpp_info.libs = ["bacnet-stack"]
```

File: recipes/bacnet-stack/all/conanfile.py (hunk 34 above)
Comment (gocarlos) [reply recipes/bacnet-stack/all/conanfile.py]: olala, I used the spdlog recipe as a template :D 

seems like a lot of stuff changed in the mean time :D

Comment (madebr) [reply recipes/bacnet-stack/all/conanfile.py]: Hahaha, this made me racing to spdlog's recipe.
Indeed, it is a bit outdated. :smile: 
Thanks for the head-up!

**Comment by uilianries** [comment]
Does it require libcrypt?

**Comment by gocarlos** [comment]
> Does it require libcrypt?

probably not: https://github.com/bacnet-stack/bacnet-stack/issues/49#issuecomment-587933953

**Comment by uilianries** [comment]
Great!

BTW @madebr Thanks for sending a patch to the upstream! :clap:

**Comment by uilianries** [comment]
> lib/libbacnet-stack.so: undefined reference to `key_set_padding'

**Comment by gocarlos** [comment]
> > lib/libbacnet-stack.so: undefined reference to `key_set_padding'

yes, it seems like the library has issues with being build as shared library

**Comment by gocarlos** [comment]
> An unexpected error happened and has been reported. Help is on its way!

cc/ @uilianries

**Comment by uilianries** [comment]
@gocarlos seems list that patch is huge and conan hooks is complaining about. Where did you find it? Could you download instead?

**Comment by gocarlos** [comment]
This is the patch from @madebr in https://github.com/bacnet-stack/bacnet-stack/pull/54

Wanted to check if it works for conan

**Comment by uilianries** [comment]
I see, well, you can download it by the url https://patch-diff.githubusercontent.com/raw/bacnet-stack/bacnet-stack/pull/54.patch

File: recipes/bacnet-stack/all/conanfile.py [hunk 36]
```diff
@@ -0,0 +1,76 @@
...
+        cmake = self._configure_cmake()
+        cmake.install()
+        tools.rmdir(os.path.join(self.package_folder, "lib", "cmake"))
+        tools.rmdir(os.path.join(self.package_folder,
+                                 "lib", "bacnet-stack", "cmake"))
+
+    def package_info(self):
+        self.cpp_info.libs = ["bacnet-stack"]
+        if self.settings.os == "Linux":
+            self.cpp_info.system_libs = ["pthread"]
```
Comment (madebr) [inline recipes/bacnet-stack/all/conanfile.py]:
```suggestion
    def package_info(self):
        self.cpp_info.libs = ["bacnet-stack"]
        if self.settings.os == "Linux":
            self.cpp_info.system_libs = ["pthread"]
        elif self.settings.os == "Windows":
            self.cpp_info.system_libs = ["ws2_32"]
        if not self.options.shared:
            self.cpp_info.defines = ["BACNET_STACK_STATIC_DEFINE"]
```

**Comment by gocarlos** [comment]
@danimtb please review

**Comment by gocarlos** [comment]
ping

**Pull request merged**